File: 11.container-with-most-water.ts
/*
 * @lc app=leetcode id=11 lang=typescript
 *
 * [11] Container With Most Water
 */

// @lc code=start
function maxArea(height: number[]): number {
  let left = 0;
  let right = height.length - 1;
  let max = 0;
  while (left < right) {
    // The longer line should be the result
    if (height[left] < height[right]) {
      max = Math.max(max, (right - left) * height[left]);
      left++;
    } else {
      max = Math.max(max, (right - left) * height[right]);
      right--;
    }
  }
  return max;
}
// @lc code=end

File: rust/11.container-with-most-water.rs
/*
 * @lc app=leetcode id=11 lang=rust
 *
 * [11] Container With Most Water
 */

// @lc code=start
use std::cmp;

impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let mut max = 0;
        let mut left = 0;
        let mut right = height.len() - 1;
        while left < right {
            if height[right] < height[left] {
                max = cmp::max(max, (right - left) as i32 * height[right]);
                right -= 1;
            } else {
                max = cmp::max(max, (right - left) as i32 * height[left]);
                left += 1;
            }
        }
        max
    }
}
// @lc code=end

struct Solution {}
